### PR TITLE
Rgw master fix plus

### PR DESCRIPTION
**Pull request by mdw-at-linuxbox**
This is actually a "jumbo" patch of 3 different fixes.  I'm combining them here just to simplify the testing/review process.

1. rgw: replace '+'.  This is an evolution of PR#16116
2. fix zero sized chunk uploads.  New fix.
3. fix small sized chunk uploads.  New fix.

With a ceph built with these fixes, I was able to get "hadoop 2.8.1" s3a tests to pass.

### CODE REVIEW

File: src/rgw/rgw_auth_s3.cc [hunk 1]
```diff
@@ -501,15 +501,22 @@ static inline std::string aws4_uri_recode(const boost::string_view& src)
...
-  if (info.request_params.empty()) {
+  const std::string *params = &info.request_params;
+  std::string copy_params;
+  if (params->empty()) {
     /* Optimize the typical flow. */
     return std::string();
   }
+  if (params->find_first_of('+') != std::string::npos) {
+    copy_params = *params;
+    boost::replace_all(copy_params, "+", " ");
```
Comment (mattbenjamin) [inline src/rgw/rgw_auth_s3.cc]: ack

Comment (ZVampirEM77) [inline src/rgw/rgw_auth_s3.cc]: Is it all right in here to replace the "+" to " "? Why not "20%"?

**Comment by rzarzynski** [comment]
jenkins retest this please (`Makefile failed to find target 'tests'`)

File: src/rgw/rgw_auth_s3.cc (hunk 1 above)
Comment (mdw-at-linuxbox) [inline src/rgw/rgw_auth_s3.cc]: I believe there is later logic that translates " " into %20 where necessary.

File: src/rgw/rgw_auth_s3.cc [hunk 2]
```diff
@@ -993,8 +993,11 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
                       std::begin(parsing_buf) + consumed);
   }
 
+  size_t stream_pos_was = stream_pos - parsing_buf.size();
+
   size_t to_extract = \
-    std::min(chunk_meta.get_data_size(stream_pos), buf_max);
+    std::min(chunk_meta.get_data_size(stream_pos_was), buf_max);
+  dout(30) << "AWSv4ComplMulti: stream_pos_was=" << stream_pos_was << ", to_extract=" << to_extract << dendl;
```
Comment (mdw-at-linuxbox) [inline src/rgw/rgw_auth_s3.cc]: The last debug comment in the function used "dout".  I thought it best to stick with the prevailing style.  A better fix would be a separate patch to handle dout/ldout.

Comment (rzarzynski) [inline src/rgw/rgw_auth_s3.cc]: Sure, this is cosmetics only. Separated patch/pull request works for me.

**Comment by mattbenjamin** [comment]
jenkins retest this please

**Comment by theanalyst** [comment]
jenkins test docs

File: src/rgw/rgw_auth_s3.cc (hunk 1 above)
Comment (zhangsw) [reply src/rgw/rgw_auth_s3.cc:512]: If the params begin with space or end with space, these spaces will be trimed in parse_key_value. 403 error still exists. Maybe we should replace "+" with "%20".

**Comment by tchaikov** [comment]
@mattbenjamin we might want to have more descriptive title in future, as (i think) it will be included as part of the release notes in Pacific.
